src/manifolds/HyperbolicHyperboloid.jl

```diff
@@ -0,0 +1,291 @@
+function check_manifold_point(M::Hyperbolic, p; kwargs...)
+    mpv =
+        invoke(check_manifold_point, Tuple{supertype(typeof(M)),typeof(p)}, M, p; kwargs...)
+    mpv === nothing || return mpv
+    if !isapprox(minkowski_metric(p, p), -1.0; kwargs...)
+        return DomainError(
+            minkowski_metric(p, p),
+            "The point $(p) does not lie on $(M) since its Minkowski inner product is not -1.",
+        )
+    end
+    return nothing
+end
+function check_manifold_point(M::Hyperbolic, p::HyperboloidPoint; kwargs...)
+    return check_manifold_point(M, p.value; kwargs...)
+end
+
+function check_tangent_vector(M::Hyperbolic, p, X; check_base_point = true, kwargs...)
+    if check_base_point
+        mpe = check_manifold_point(M, p; kwargs...)
+        mpe === nothing || return mpe
+    end
+    mpv = invoke(
+        check_tangent_vector,
+        Tuple{supertype(typeof(M)),typeof(p),typeof(X)},
+        M,
+        p,
+        X;
+        check_base_point = false, # already checked above
+        kwargs...,
+    )
+    mpv === nothing || return mpv
+    if !isapprox(minkowski_metric(p, X), 0.0; kwargs...)
+        return DomainError(
+            abs(minkowski_metric(p, X)),
+            "The vector $(X) is not a tangent vector to $(p) on $(M), since it is not orthogonal (with respect to the Minkowski inner product) in the embedding.",
+        )
+    end
+    return nothing
+end
+function check_tangent_vector(
+    M::Hyperbolic,
+    p::HyperboloidPoint,
+    X::HyperboloidTVector;
+    kwargs...,
+)
+    return check_tangent_vector(M, p.value, X.value; kwargs...)
+end
+
+function convert(::Type{HyperboloidTVector}, X::T) where {T<:AbstractVector}
+    return HyperboloidTVector(X)
+end
+function convert(::Type{HyperboloidTVector}, (p, X)::Tuple{T,T}) where {T<:AbstractVector}
```

Comment:
What do you need this method for?

Reply:
Classically for conversion of tangents `X` you compute the push forward \pi_*(p)[X]. Thats why you need `(p,X)`as input for tangent vector conversions. To have this unifies between all 4 types (Array, Hyperboloid, PoincareBall, PoincareHalfplane), I implemented conversions from and to any of these 4 types. Yes, this is just wrapping an array in the HyperboloidTVector – for in this case we do not need `p`, but in the general interface we do.

Reply:
To be more specific, why is the second argument a tuple?

Reply:
The first (`p`) is a point, the second (`X`) is the tangent vector. Since Arrays (Vectors to be precise) representing our points/vectors do not have types, its just a plain tuple of two arrays.

Reply:
Sorry, I didn't see your response when I wrote my second comment. I'll think about this converting.

Reply:
Could you change the signature here to something like `convert(::Hyperbolic, ::Type{HyperboloidTVector}, p, X)`? We don't have to necessarily take two arguments in `convert`. This method as written is incompatible with typical usage of two-argument `convert` anyway.

Reply:
Why is It incompatible? Because the Tuple introduces too many ambiguities?
For sure we can `de-Tuple` this, since just with `X` it does not work anyways, we are breaking a little bit anyways;

So as long as we agree on a unified way, I am also fine with `p, X` instead of `(p,X)`, both are fine to dispatch on for me. And you'Re right the first avoids the ambiguity with several `Tuple` converts

Reply:
Should I then also remove all these combined ones? I tried to then also do a combined “convert both p and X” methods here, since `p` is converted anyways.

Reply:
> Why is It incompatible? Because the Tuple introduces too many ambiguities?

Primarily because someone who wants to use that methods has to make that tuple anyway. Such tuples don't naturally occur in Manifolds.jl. Technically there shouldn't be any ambiguities (in the multiple dispatch sense) I think.

Reply:
> Should I then also remove all these combined ones? I tried to then also do a combined “convert both p and X” methods here, since `p` is converted anyways.

Combined ones would be fine I think but in a slightly different format:
```
convert(::Type{Tuple{HyperboloidPoint,HyperboloidTVector}}, (p, X)::Tuple{T,T}) where {T<:AbstractVector}
```
If both the argument you convert *and* and type you convert to are tuples then it's also fine to return a tuple I think.

Reply:
So I think it's fine to have tuples in all three places (type, converted argument, returned object) or in none.

Reply:
Din line 160 I have
````julia
function convert(
     ::Type{Tuple{HyperboloidPoint,HyperboloidTVector}},
     (p, X)::Tuple{PoincareBallPoint,PoincareBallTVector},
 )
````
well internally it called the “just `X`” one – but that one is ok then (similarly for a tuple fo arrays of course)?

That would mean, that a you can never `convert` to just a vector, right?
Because if you convert `(p,X)` to `(q,Y)` converting `X` might require `p` but only rarely requires `q` (on Hyperbolic it never requires `q`).
 So if you just want to convert a tangent, oi.e. get `Y` from `(p,X)` one can spare the computation to convert `p` to `q`. That's my intention of this method.

But sure we can keep the tuple ones we have and do a `convert(::HyperboloidTVector, p, X)`for that, or not?

Reply:
> well internally it called the “just `X`” one – but that one is ok then (similarly for a tuple fo arrays of course)?

Yes, this looks fine.

> That would mean, that a you can never `convert` to just a vector, right?

With a 2-argument `convert` that requires the point? No, we could not. I think for the case of converting just the tangent vector when the point is also required we should use a three-argument `convert` like you wrote in the last sentence of your message.

> Because if you convert `(p,X)` to `(q,Y)` converting `X` might require `p` but only rarely requires `q` (on Hyperbolic it never requires `q`).

In which case `q` isn't uniquely defined by `p`?

Reply:
Oh, its always uniquely determined by `p`, sorry if I wrote that wrong. We can just omit the computation to _compute_ `q` (memory allocation and computational time) if its not needed.

And ok, I will convert the tangent vector converts to 3-argument ones. That seems to be the best.

Reply:
Thanks, this refactoring made much of the tuple code much clearer and removed 3 ambiguities the tuple-nontuple conversios introduced.

Reply:
That's great :slightly_smiling_face: .